Add surface syntax for dependent pairs

axch opened the pull request:
Fixes #966.

Co-authored-by: Dougal Maclaurin <dougalm@google.com>

apaszke commented on src/lib/Inference.hs:
```diff
@@ -964,15 +964,19 @@ checkOrInferRho (WithSrcE pos expr) reqTy = do
+      let checkNoMissing = (addSrcContext pos'
+           $ unless (null $ eSetToList missingDs) $ throw TypeErr
+           $ "Couldn't synthesize a class dictionary for: "
+             ++ pprint (head $ eSetToList missingDs))
```
Please avoid reformatting if you're not actually modifying the code. Or split the reformatting changes into a separate commit

axch replied:
Split into separate commit.

apaszke commented on tests/type-tests.dx:
```diff
@@ -350,13 +350,16 @@ def triRefIndex {h n} (ref:Ref h ((i':n)=>(..i')=>Float)) (i:n) : Ref h ((..i)=>
--- input type with a non-dependent function type, leading to a later unification errors.
+-- There was a time when this wasn't possible, because checking mode
+-- would unify the input type with a non-dependent function type,
+-- leading to a later unification errors.
```
Again, please don't format things unnecessarily

apaszke commented on src/lib/Inference.hs:
```diff
@@ -1008,6 +1012,26 @@ checkOrInferRho (WithSrcE pos expr) reqTy = do
+        cheapReduceWithDecls decls result >>= \case
+          (Just rhs', DictTypeHoistSuccess, []) -> return rhs'
+          _ -> throw TypeErr $ "Can't reduce type expression: " ++
+                 docAsStr (prettyBlock decls result)
```
Is this any different from `checkUType`?

apaszke reviewed (commented):
Can you add some more tests that actually evaluate some expressions using dependent pairs? Perhaps have a case statement that returns lists of different lengths? And make sure we can print those.

axch commented:
Also added a test the involves going through a case, and printing out a dependent pair.  Added a clause for dependent pairs to `Serialize.hs` to make them print more nicely than the default.

apaszke commented on tests/type-tests.dx:
```diff
@@ -520,6 +520,24 @@ foo : Fin 3 => Float = [1.0, 2.0, 3.0]
+if False
+  then (5 ,> buffer) :: tp
+  else (3 ,> for i:(Fin 3). (ordinal i) * 2) :: tp
+> (3 ,> [0, 2, 4]) @ (n:Nat &> ((Fin n) => Nat))
```
The `@` printing is a little unfortunate, since it is an operator we use for index casting. Perhaps use the `::` syntax?

In the long term, we should consider replacing `::` with `@`, since that might be less confusing. We would need to find another operator for index casts, though.

axch replied:
Done.
